src/WebJobs.Script/Host/FunctionMetadataManager.cs

```diff
@@ -89,5 +150,56 @@ internal bool IsScriptFileDetermined(FunctionMetadata functionMetadata)
             }
             return true;
         }
+
+        private void LoadCustomProvidersFunctions(List<FunctionMetadata> functionMetadataList)
+        {
+            if (_functionProviders != null && _functionProviders.Any())
+            {
+                AddMetadataFromCustomProviders(functionMetadataList);
+                AddErrorsFromCustomProviders();
+            }
+        }
+
+        private void AddMetadataFromCustomProviders(List<FunctionMetadata> functionMetadataList)
+        {
+            var functionProviderTasks = new List<Task<ImmutableArray<FunctionMetadata>>>();
+            foreach (var functionProvider in _functionProviders)
+            {
+                functionProviderTasks.Add(functionProvider.GetFunctionMetadataAsync());
+            }
+
+            var functionMetadataListArray = Task.WhenAll(functionProviderTasks).GetAwaiter().GetResult();
+
+            // This is used to make sure no duplicates are registered
+            var distinctFunctionNames = new HashSet<string>(functionMetadataList.Select(m => m.Name));
+
+            foreach (var metadataArray in functionMetadataListArray)
+            {
+                if (!metadataArray.IsDefaultOrEmpty)
+                {
+                    foreach (var metadata in metadataArray)
+                    {
+                        if (distinctFunctionNames.Contains(metadata.Name))
+                        {
+                            throw new InvalidOperationException($"Found duplicate {nameof(FunctionMetadata)} with the name {metadata.Name}");
+                        }
+
+                        // All custom provided functions are considered codeless functions
+                        metadata.IsCodeless = true;
+
+                        distinctFunctionNames.Add(metadata.Name);
+                        functionMetadataList.Add(metadata);
+                    }
+                }
+            }
+        }
+
+        private void AddErrorsFromCustomProviders()
+        {
+            foreach (var provider in _functionProviders)
+            {
+                provider.FunctionErrors?.ToList().ForEach(kvp => _functionErrors[kvp.Key] = kvp.Value.ToList());
```

Comment:
nit: just do a foreach (var error in provider.FunctionErrors), so you don't need the ToList()